src/Compilers/CSharp/Portable/Symbols/SubstitutedFieldSymbol.cs

```diff
@@ -106,13 +106,17 @@ public override bool Equals(Symbol obj, TypeCompareKind compareKind)
                 return true;
             }
 
-            var other = obj as SubstitutedFieldSymbol;
-            return (object)other != null && TypeSymbol.Equals(_containingType, other._containingType, compareKind) && OriginalDefinition == other.OriginalDefinition;
+            var other = obj as FieldSymbol;
+            return (object)other != null && TypeSymbol.Equals(_containingType, other.ContainingType, compareKind) && OriginalDefinition == other.OriginalDefinition;
         }
 
         public override int GetHashCode()
         {
-            return Hash.Combine(_containingType, OriginalDefinition.GetHashCode());
+            // Substituted fields can still be considered equal to other field symbols when 
+            // the substituted type is considered equal (e.g. when ignoring nullability).
+            // We defer the hashcode to the original definition so it will match in those scenarios. 
+            // See https://github.com/dotnet/roslyn/issues/38195
+            return this.OriginalDefinition.GetHashCode();
```

Comment:
>this.OriginalDefinition.GetHashCode() [](start = 19, length = 37)

I think it is still valid to return ```Hash.Combine(_containingType, OriginalDefinition.GetHashCode());```, unless ```TypeSymbol.Equals(_containingType, other.ContainingType, ALL_IGNORE_OPTIONS)``` is true

Reply:
For performance, we can replace this condition for containing types with comparison of hash codes

---
In reply to: [370326121](https://github.com/dotnet/roslyn/pull/41123#discussion_r370326121) [](ancestors = 370326121)

Reply:
But how would we know that in the hashcode function? We wouldn't have an other to compare against? If we put a substituted method symbol into a dictionary, then did a .Contains() with a regular symbol that would compare equal when ignoring nullability we would fail, even with a comparer that accounted for it.

Reply:
I suppose we could make a hashcode variant that accepts the TypeCompareKind? Then by default we'd use ignore nullabillity?

Reply:
To clarify, the ```other``` in the condition is ```this.OriginalDefinition```

---
In reply to: [370326741](https://github.com/dotnet/roslyn/pull/41123#discussion_r370326741) [](ancestors = 370326741,370326121)